File: app/src/main/java/org/monksanctum/xand11/graphics/GraphicsManager.java
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.monksanctum.xand11.graphics;

import android.util.SparseArray;

import org.monksanctum.xand11.XService;
import org.monksanctum.xand11.errors.DrawableError;
import org.monksanctum.xand11.errors.GContextError;
import org.monksanctum.xand11.errors.PixmapError;
import org.monksanctum.xand11.errors.ValueError;

public class GraphicsManager {

    public static final boolean DEBUG = XService.GRAPHICS_DEBUG;

    private final SparseArray<GraphicsContext> mGcs = new SparseArray<>();
    private final SparseArray<Pixmap> mPixmaps = new SparseArray<>();
    private final SparseArray<XDrawable> mDrawableLookup = new SparseArray<>();

    private final ColorMaps mColorMaps = new ColorMaps();

    public ColorMaps getColorMap() {
        return mColorMaps;
    }

    public void addDrawable(int id, XDrawable drawable) {
        synchronized (mDrawableLookup) {
            mDrawableLookup.put(id, drawable);
        }
    }

    public void removeDrawable(int id) {
        synchronized (mDrawableLookup) {
            mDrawableLookup.remove(id);
        }
    }

    public XDrawable getDrawable(int id) throws DrawableError {
        synchronized (mDrawableLookup) {
            XDrawable drawable = mDrawableLookup.get(id);
            if (drawable == null) {
                throw new DrawableError(id);
            }
            return drawable;
        }
    }

    public GraphicsContext createGc(int id) {
        GraphicsContext graphicsContext = new GraphicsContext(id);
        synchronized (mGcs) {
            mGcs.put(id, graphicsContext);
        }
        return graphicsContext;
    }

    public GraphicsContext getGc(int id) throws GContextError {
        synchronized (mGcs) {
            GraphicsContext graphicsContext = mGcs.get(id);
            if (graphicsContext == null) {
                throw new GContextError(id);
            }
            return graphicsContext;
        }
    }

    public void freeGc(int id) {
        synchronized (mGcs) {
            mGcs.remove(id);
        }
    }

    public Pixmap createPixmap(int id, byte depth, int width, int height, XDrawable drawable)
            throws ValueError {
        Pixmap pixmap = new Pixmap(depth, width, height, id, drawable);
        synchronized (mPixmaps) {
            mPixmaps.put(id, pixmap);
        }
        addDrawable(id, pixmap);
        return pixmap;
    }

    public void freePixmap(int id) {
        synchronized (mPixmaps) {
            mPixmaps.remove(id);
        }
        removeDrawable(id);
    }

    public Pixmap getPixmap(int pixmap) throws PixmapError {
        synchronized (mPixmaps) {
            Pixmap p = mPixmaps.get(pixmap);
            if (p == null) {
                throw new PixmapError(pixmap);
            }
            return p;
        }
    }
}

File: core/src/main/java/com/monksanctum/xand11/core/graphics/GraphicsManager.kt
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.monksanctum.xand11.graphics


import org.monksanctum.xand11.core.GRAPHICS_DEBUG
import org.monksanctum.xand11.core.GraphicsContext
import org.monksanctum.xand11.core.Throws
import org.monksanctum.xand11.errors.DrawableError
import org.monksanctum.xand11.errors.GContextError
import org.monksanctum.xand11.errors.PixmapError
import org.monksanctum.xand11.errors.ValueError

class GraphicsManager {

    private val mGcs = mutableMapOf<Int, GraphicsContext>()
    private val mPixmaps = mutableMapOf<Int, Pixmap>()
    private val mDrawableLookup = mutableMapOf<Int, XDrawable>()

    val colorMap = ColorMaps()

    fun addDrawable(id: Int, drawable: XDrawable) {
        synchronized(mDrawableLookup) {
            mDrawableLookup.put(id, drawable)
        }
    }

    fun removeDrawable(id: Int) {
        synchronized(mDrawableLookup) {
            mDrawableLookup.remove(id)
        }
    }

    @Throws(DrawableError::class)
    fun getDrawable(id: Int): XDrawable {
        synchronized(mDrawableLookup) {
            return mDrawableLookup.get(id) ?: throw DrawableError(id)
        }
    }

    fun createGc(id: Int): GraphicsContext {
        val graphicsContext = GraphicsContext(id)
        synchronized(mGcs) {
            mGcs.put(id, graphicsContext)
        }
        return graphicsContext
    }

    @Throws(GContextError::class)
    fun getGc(id: Int): GraphicsContext {
        synchronized(mGcs) {
            return mGcs.get(id) ?: throw GContextError(id)
        }
    }

    fun freeGc(id: Int) {
        synchronized(mGcs) {
            mGcs.remove(id)
        }
    }

    @Throws(ValueError::class)
    fun createPixmap(id: Int, depth: Byte, width: Int, height: Int, drawable: XDrawable): Pixmap {
        val pixmap = Pixmap(depth, width, height, id, drawable)
        synchronized(mPixmaps) {
            mPixmaps.put(id, pixmap)
        }
        addDrawable(id, pixmap)
        return pixmap
    }

    fun freePixmap(id: Int) {
        synchronized(mPixmaps) {
            mPixmaps.remove(id)
        }
        removeDrawable(id)
    }

    @Throws(PixmapError::class)
    fun getPixmap(pixmap: Int): Pixmap {
        synchronized(mPixmaps) {
            return mPixmaps.get(pixmap) ?: throw PixmapError(pixmap)
        }
    }

    companion object {

        val DEBUG = GRAPHICS_DEBUG
    }
}
